### PR TITLE
add hooks to invoke callbacks when child windows closed

### LINKED ISSUE #7: Add hook which closes parent window whenever all child windows close

### LINKED ISSUE #8: add hook which closes parent window on child window close

### PR DESCRIPTION
**Pull request by jlight-scottlogic**
Resolves #7 
Resolves #8 

Rather than hardcoding the "close parent window" functionality into the hook, I allowed a callback to be passed so that different behaviour can be invoked as a result of either one/all of the child windows being closed.

### CODE REVIEW

File: demo/src/pages/callbackWhenAllChildrenClosed/CallbackWhenAllChildrenClosed.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    const mainWindow = window.fin.Window.getCurrentSync();
+
+    useCallbackWhenAllChildWindowsClosed(
+        (parent: _Window) => alert(\`All children of window \${parent.identity.name} have been closed\`),
+        mainWindow
+    );
+
+    return (
+        <div></div>
```
[the remark this answers is not in the document]
Comment (jlight-scottlogic) [reply demo/src/pages/callbackWhenAllChildrenClosed/CallbackWhenAllChildrenClosed.tsx]: this is just the example text (ie. this is only part of a string variable and the code doesn't actually exist). do you still think it's worth changing?

**Comment by jlight-scottlogic** [comment]
> The version in the package.json and package-lock.json should be bumped to 1.4.0. Apart from that, it looks good :)

I wasn't sure whether we should just do one major version bump after all of these changes go in. If your work is changing any signatures of any methods in the existing API, that is. Happy to add a minor version bump either way.

File: demo/src/pages/callbackWhenAllChildrenClosed/CallbackWhenAllChildrenClosed.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ms14981) [reply demo/src/pages/callbackWhenAllChildrenClosed/CallbackWhenAllChildrenClosed.tsx]: Totally up to you, it doesn't make much of a difference either way :)

[the remark this answers is not in the document]
Comment (jlight-scottlogic) [reply demo/src/pages/callbackWhenAllChildrenClosed/CallbackWhenAllChildrenClosed.tsx]: returned null instead as we don't care about the view

File: demo/src/pages/callbackWhenAnyChildClosed/CallbackWhenAnyChild.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    const mainWindow = window.fin.Window.getCurrentSync();
+
+    useCallbackWhenAnyChildWindowClosed(
+        (parent, child) =>
+            alert(\`A child window (\${child.identity.name}) of window \${parent.identity.name} has been closed\`),
+        mainWindow,
+    );
+
+    return (
+        <div></div>
```
[the remark this answers is not in the document]
Comment (jlight-scottlogic) [reply demo/src/pages/callbackWhenAnyChildClosed/CallbackWhenAnyChild.tsx]: returned null here as well

File: package.json.orig [hunk 3]
```diff
@@ -0,0 +1,52 @@
+{
```
Comment (oriondean) [inline package.json.orig]: Looks like a conflict related file has been committed?

**Pull request merged**